File: solutions/typescript/robot-name/1/robot-name.ts
const NAME_REGISTER = new Set<string>();

export default class RobotName {
  public name: string;

  constructor() {
    this.name = this.generateName();
  }

  private randomLetter() {
    const num = Math.floor(Math.random() * 26 + 65);

    return String.fromCharCode(num);
  }

  private randomNumber() {
    return Math.floor(Math.random() * 9);
  }

  private generateName(): string {
    const letters = [
      this.randomLetter(),
      this.randomLetter(),
      this.randomNumber(),
      this.randomNumber(),
      this.randomNumber(),
    ];

    let name = letters.join("");

    if (NAME_REGISTER.has(name)) {
      name = this.generateName();
    }

    NAME_REGISTER.add(name);

    return name;
  }

  public resetName() {
    this.name = this.generateName();
  }
}

File: solutions/csharp/robot-name/1/RobotName.cs
using System.Text;

public class Robot
{

    public Robot() => Reset();

    private readonly Random _rng = new ();
    
    private static readonly HashSet<string> UsedNamed = [];
    
    public string Name { get; private set; }

    public void Reset() => Name = GenerateName();

    private string GenerateName()
    {
        var name = GetRandomName();

        while (!UsedNamed.Add(name))
        {
            name = GetRandomName();
        }
        
        return name;
    }

    private string GetRandomName() => $"{RandomLetter()}{RandomLetter()}{RandomNumber()}{RandomNumber()}{RandomNumber()}";

    private char RandomLetter() => Convert.ToChar('A' + _rng.Next(0, 26));

    private string RandomNumber() => _rng.Next(0, 10).ToString();
}